pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,363 @@
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		for _, l := range lbls {
+			if l.Name == cal {
+				labelValues[idx] = l.Value
+				break
+			}
+		}
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	return labelValues
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		labelValues[idx] = lbls.Get(cal)
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	key := t.hashLabelValues(labelValues)
+	t.updateObservations(key, ts.Unix(), activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+	t.updateState(ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement)
+}
+
+// updateObservations updates or creates a new observation in the 'observed' map.
+func (t *Tracker) updateObservations(key string, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	t.observedMtx.RLock()
+	o, known := t.observed[key]
+	t.observedMtx.RUnlock()
+
+	if !known {
+		if len(t.observed) < t.maxCardinality*2 && createIfDoesNotExist {
+			// When createIfDoesNotExist is false, it means that the method is called from DecrementActiveSeries, when key doesn't exist we should ignore the call
+			// Otherwise create a new observation for the key
+			t.createNewObservation(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason)
+		}
+		return
+	}
+
+	o.lastUpdate.Store(ts)
+	if activeSeriesIncrement != 0 {
+		o.activeSerie.Add(activeSeriesIncrement)
+	}
+	if receivedSampleIncrement > 0 {
+		o.receivedSample.Add(receivedSampleIncrement)
+	}
+	if discardedSampleIncrement > 0 && reason != nil {
+		o.discardedSampleMtx.Lock()
+		if _, ok := o.discardedSample[*reason]; ok {
+			o.discardedSample[*reason].Add(discardedSampleIncrement)
+		} else {
+			o.discardedSample[*reason] = atomic.NewFloat64(discardedSampleIncrement)
+		}
+		o.discardedSampleMtx.Unlock()
+	}
+}
+
+// updateState checks if the tracker has exceeded its max cardinality and updates overflow state if necessary.
+func (t *Tracker) updateState(ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64) {
+	previousOverflow := true
+	t.observedMtx.RLock()
+	// Transition to overflow mode if maximum cardinality is exceeded.
+	if !t.isOverflow.Load() && len(t.observed) > t.maxCardinality {
+		// Make sure that we count current overflow only when state is switched to overflow from normal.
+		previousOverflow = t.isOverflow.Swap(true)
+		if !previousOverflow {
+			// Initialize the overflow counter.
+			t.overflowCounter = &observation{}
```

Comment:
There should be some kind of concurrency coordination here on setting this property.

Reply:
this logic is gone in f28d67212d212e1c43b623a47ec887b09d679f82